scene/animation/animation_node_state_machine.h

```diff
@@ -116,6 +116,11 @@ class AnimationNodeStateMachine : public AnimationRootNode {
 		STATE_MACHINE_TYPE_GROUPED,
 	};
 
+	enum {
+		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 6,
+		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED = 7
+	};
+
```

Comment:
```suggestion
	enum {
		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 6,
		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED = 7
	};
```
Managing IDs with sequential numbers across classes is not recommended.

If you really want to do this, at least for example, have the ID for each AnimationNode in an enum on the AnimationTree side, and multiply it by 1000 to add it, so that the AnimationNode can have signals from 0 to 999.

However, I think it is still complicated, so I think it is easiest for the users to understand by passing class names to the signals.

Reply:
I mean at that point would it be best to just have the full enum defined on the base `AnimationNode`? The per signal documentation wouldn't be on the specific node that emits it but at least it would all be in one place and the values would be unique. I think it would be very likely for the user to make the mistake of comparing against a normal string, and it feels weird passing the parameter as a type when it's already a part of how the enum is defined (currently twice).

Reply:
```gdscript
if node_type == &"AnimationNodeStateMachine" and what == AnimationNodeStateMachine.ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED:
```
Yes, this is recommended with
```
ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::STRING_NAME, "animation_node_class_name"), PropertyInfo(Variant::INT, "what")));
```

Reply:
Were there issues/regrets with the notification guidance of `You can make your own, but don't use the same numbers as other notifications in other nodes`? I don't think having two parameters here makes it per se safer when the enum values are allowed to overlap compared to if the enum values were not allowed to overlap to begin with. While there is some overhead to checking existing values (ie search the codebase for `ANIMATION_NODE_NOTIFICATION_[A-Z_]+ = \d+`) I don't think it's too out of question for how infrequently it would happen in this case.

I can definitely just give each class a range of 100 or 1000 values to make it easier though. If that "assigned" range ever ran out adding new values without changing old ones would still possible since values don't need to be contiguous, but probably it would be worth figuring out something different entirely for handling whatever would cause that.

Reply:
First, I think we can agree on the following two points.

- Signals should be defined for each AnimationNode class
- The number should not be a sequential number across classes

Therefore, overlapping numbers per class, e.g., `ANIMATION_NODE_NOTIFICATION_ONESHOT_STARTED = 0` and `ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 0`,  is not a problem.

However, if `animation_node_notification()` only returns int and pathname, there is a problem with users cannot distinguish between them. So in order to solve that problem, animation_node_notification must return the class_name.

Reply:
I agree with the first point but not the second. I think having repeat notification values and relying on an extra path to be passed is error prone compared to making the enum values distinct across classes, although at least with the current 8 values "started" and "finished" would always be 0 and 1 so if someone tab completed the wrong one of those it would still work (as long as it was also qualified by the type or path). That's why I was asking if more generally there was a consensus about how notifications are split across classes being bad, since that's the closest inspiration for how the signal is defined and used at the moment. It took me only a few seconds to find all the unique `NOTIFICATION_` definitions so I don't think it's too unreasonable for animation notifications to have distinct values.

Reply:
The second problem, as raised earlier, is that the sequential numbering is broken when signals are added or deleted. This is unacceptable.

The long enum names are unavoidable in order to conform to Godot's naming conventions, and this is not the only area where long enums are used.

Reply:
> The second problem, as raised earlier, is that the sequential numbering is broken when signals are added or deleted. This is unacceptable.

By sequential numbering do you mean specifically having the enum values 0, 1, 2 etc for each class without gaps? In my experience when constants like these are deprecated/removed, a hole is left in the numbering so that serialized data still maps to the correct name. For adding enums, I do think it's best to allocate a large range per class (so its enum values wouldn't have gaps, but also wouldn't start at 0) if mostly for convenience, but even without that I don't think there's any technical problem with having gaps in a given class's notification values (`Node` does this, for example).

> I don't know what you mean. The type-safe check is only done once by comparing class name StringName

I should have been clearer. You're right the `node_type` is the part actually doing the type check in that case but I was talking about literally specifying the type. For example in the following we want to check that a state machine has finished:
```gdscript
if node_type == &"AnimationNodeStateMachine" and what == AnimationNodeStateMachine.ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED:
```
The type gets specified in the first equality, the classname itself of course, and then the enum name. Having the name as a separate parameter would be useful if we wanted to define the enum on a common base class or wanted each class to be zero-indexed but I don't think either is correct/necessary here. IMO, it doesn't need to be in the enum name because it always is qualified by the class preceding it, but I don't feel as strongly about this.

I'd prefer to cut out the name there though than to remove `ANIMATION_NODE_` as it might make people think it gets used with `_notification` (also sort of the reason I originally went with a different word entirely, although that too does not scale well in a large codebase). Maybe `AnimationNodeStateMachine.ANIMATION_EVENT_STARTED` would be better if it was acceptable to use a different word than notification.